### PR TITLE
Redundant styles

### PR DESCRIPTION
**Pull request by stevenbfine**
Remove redundant style properties and values in style attributes.

### CODE REVIEW

File: popup.js [hunk 1]
```diff
@@ -67,18 +69,94 @@ function outputHTMLString(messages) {
...
+function fillRemainingHolesAndMinimizeStyles(messages, i) {
   var html = messages[i].html;
   var frameHoles = messages[i].frameHoles;
   for (var index in frameHoles) {
     if (frameHoles.hasOwnProperty(index)) {
       var frameIndex = frameHoles[index];
       for (var j = 0; j < messages.length; j++) {
         if (messages[j].frameIndex == frameIndex) {
-          fillRemainingHoles(messages, j);
+          fillRemainingHolesAndMinimizeStyles(messages, j);
```
Comment (progers) [inline popup.js:80]: Does this miss the head and html elements? Looking at the output of dglazkov.com, there still seems to be a few large blocks of redundant styles.

File: tests/tests.js [hunk 2]
```diff
@@ -364,59 +364,55 @@ QUnit.test('processText: nested escaped characters', function(assert) {
...
+});
+
+QUnit.test('processTree: element with id', function(assert) {
+  var serializer = new HTMLSerializer();
+  var fixture = document.getElementById('qunit-fixture');
+  var element = document.createElement('div');
+  element.setAttribute('id', 'myId');
+  fixture.appendChild(element);
+  serializer.processTree(element);
+  assert.equal(serializer.html[0], '<div ');
```
Comment (progers) [inline tests/tests.js:399]: Now that the serialized result is small, can you update tests (or add a test) that does something like:
assert.equal(serializer.finalResult, "`<div id='myId'></div>`");

File: popup.js (hunk 1 above)
Comment (stevenbfine) [inline popup.js:80]: Yes, it does miss those elements.  It's because document.getElementById isn't returning them.  For some reason, even though their ids are specified in the html, the document that is rendered does not have ids for those elements.  I think it might have to do with where I place the style elements (I place the style tags for pseudo elements and fonts before the html and head elements), but I am looking into it.

File: tests/tests.js (hunk 2 above)
Comment (stevenbfine) [inline tests/tests.js:399]: I don't think I'd be able to test for this with a unit test, because HTMLSerializer still produces a document with large styles.  It's only in popup.js that the styles are minimized.  Should I try to write a simple test that goes through the whole process of using HTMLSerializer and then popup?

Comment (progers) [inline tests/tests.js:399]: Ah, good idea, an end-to-end test sounds perfect. Yeah, please do that instead. It will show that the styles really are minimized, for example.

Comment (stevenbfine) [inline tests/tests.js:399]: Just finished the test.  Turned out to be a bit more intricate than I had anticipated.

File: popup.js (hunk 1 above)
Comment (stevenbfine) [inline popup.js:80]: It didn't work for the head element because there are style tags that are placed above the head tag (we discussed this a few weeks ago with Chris, and chose to place the style tags that contain fonts and pseudo elements at the top for simplicity).  Therefore when the document is rendered a new head element is created.  I just pushed a fix (writing a test now).

It doesn't work for the html node because setting iframe.contentDocument.documentElement.innerHTML produces a document with a new html element.  Perhaps I should process the html node separately?

It also doesn't work for pseudo elements.  It wouldn't be a simple fix, but definitely doable.  Do you think that's high on the priority list?

Comment (progers) [inline popup.js:80]: I think it's okay to skip the pseudo element stuff for now.

Could you process the html element though? Without that, a simple hello world webpage still has a huge block of redundant style.

File: popup.js [hunk 3]
```diff
@@ -67,18 +69,94 @@ function outputHTMLString(messages) {
...
+    var fullStyleDeclaration = message.html[index];
+    var element = doc.getElementById(id);
+    if (element) {
+      element.removeAttribute('style');
+      var unstyledStyle = doc.defaultView.getComputedStyle(element, null);
+      var stylePrefix;
+      var styleSuffix;
+      // Remove style declaration from style attribute.
+      var style = fullStyleDeclaration.replace(
+        /^style=("|&(amp;)*?quot;)(.*)("|&(amp;)*?quot;) ?$/,
```
Comment (progers) [inline popup.js]: This regex scares me, and I worry that re-parsing the style (here and below) could lead to bugs if the regex isn't perfect. WDYT of using maps instead so we don't need to do this parsing?

Something like this:

```
    if (element) {
      var originalStyle = element.getAttribute('style');
      var styledStyle = doc.defaultView.getComputedStyle(element, null);

      // We need to save off the CSSComputedStyle values because this object is
      // 'live' and removing the style attribute will affect it.
      var styledStyleMap = {};
      for (var propertyIndex = 0; propertyIndex < styledStyle.length; propertyIndex++) {
        var propertyName = styledStyle.item(propertyIndex);
        styledStyleMap[propertyName] = styledStyle.getPropertyValue(propertyName);
      }

      element.removeAttribute('style');
      var unstyledStyle = doc.defaultView.getComputedStyle(element, null);
      var neededStyles = '';
      for (var propertyName in styledStyleMap) {
        if (styledStyleMap[propertyName] !== unstyledStyle.getPropertyValue(propertyName))
          neededStyles += `${propertyName}: ${styledStyleMap[propertyName]}; `;
      }

      // FIXME: maybe expose escapedCharacter as a static function?
      // FIXME: this doesn't work with the extension because html serializer is not defined. Maybe import it?
      var escapedStyleQuotes = (new HTMLSerializer()).escapedCharacter('"', nestingDepth+1);
      neededStyles = neededStyles.replace(/"/g, escapedStyleQuotes);
      var quote = (new HTMLSerializer()).escapedCharacter('"', nestingDepth);
      message.html[index] = `style=${quote}${neededStyles}${quote} `;
      element.setAttribute('style', originalStyle);
    }
```

File: popup.js [hunk 4]
```diff
@@ -67,18 +69,94 @@ function outputHTMLString(messages) {
...
+          unstyledValue = unstyledStyle[property].replace(/"/g, '');
+          if (unstyledValue == value) {
+            return '';
+          } else {
+            return match;
+          }
+        }
+      );
+      message.html[index] = stylePrefix + style + styleSuffix;
+      element.setAttribute('style', style);
```
Comment (progers) [inline popup.js]: This could lead to cascading bugs if this function or the browser has bugs. Could you save off the original style and replace it here, instead of using the minimized style?

File: tests/tests.js [hunk 5]
```diff
@@ -536,3 +532,48 @@ QUnit.test('escapedCharacterString', function(assert) {
...
+    'idToStyleIndex': {"myId": 1},
+    'windowHeight': 5,
+    'windowWidth': 5,
+    'frameIndex': '0'
+  };
+  minimizeStyles(message);
+  assert.equal(message.html[1], 'style="width: 5px;" ');
+});
+
+QUnit.test('serialize tree: end-to-end', function(assert) {
```
Comment (progers) [inline tests/tests.js]: Can you add a second test with a non-empty style?

```
  var style = document.createElement('style');
  style.innerHTML = 'div { border: 1px solid blue; }';
  iframe.contentDocument.body.appendChild(style);
```

This will show that weird transform-origin behavior which is good to test.

File: popup.js [hunk 6]
```diff
@@ -1,5 +1,7 @@
 document.addEventListener('DOMContentLoaded', function() {
-  document.getElementById('button').addEventListener('click', click);
+  if (document.getElementById('button')) {
```
Comment (progers) [inline popup.js:2]: Is this change needed?

File: popup.js [hunk 7]
```diff
@@ -67,18 +69,94 @@ function outputHTMLString(messages) {
...
+        function(match, property, p2, p3, value) {
+          unstyledValue = unstyledStyle[property].replace(/"/g, '');
+          if (unstyledValue == value) {
+            return '';
+          } else {
+            return match;
+          }
+        }
+      );
+      message.html[index] = stylePrefix + style + styleSuffix;
```
Comment (progers) [inline popup.js]: Optimization idea: if style is empty, just set this to ""

**Comment by progers** [comment]
Added a bunch of comments but overall this looks great. I think we can land it tomorrow.

My suggestion of using a style map is a little involved and might not work... maybe we could leave the regexes and file a bug to explore that as a followup?

File: popup.js (hunk 6 above)
Comment (stevenbfine) [inline popup.js:2]: I added it because otherwise it throws an error when I import popup.js into my tests.

File: popup.js (hunk 1 above)
Comment (stevenbfine) [inline popup.js:80]: Done

File: popup.js (hunk 4 above)
Comment (stevenbfine) [inline popup.js]: Done

File: popup.js (hunk 7 above)
Comment (stevenbfine) [inline popup.js]: Done

File: tests/tests.js (hunk 5 above)
Comment (stevenbfine) [inline tests/tests.js]: I created an additional test, similar to the one above, but with those three lines added.  The resulting style attribute is relatively large.  Let me know if the test I added is what you had in mind.

Also, remind me about the transform-origin behavior?  Why does this attribute change the transform origin?  Is it because the border changes the center point of the element?

Comment (progers) [inline tests/tests.js]: Test looks great!

I was surprised to see transform-origin change when the border changes, but I think it's just how the system works... Possibly transform-origin is spec'd to be the center of an object by default?

**Comment by progers** [comment]
This is looking great and dglazkov.com minimizes nicely now. Is this ready to roll?

**Comment by stevenbfine** [comment]
Thanks! I think I've addressed everything other than the style map.  Should I work on that, of file a bug for now, and possible come back to it?

**Comment by progers** [comment]
Want to investigate that as a followup? LGTM if you'd like to land as-is, or looking into that style map is okay too.

**Comment by stevenbfine** [comment]
I think for now, I'll leave it and file it as a bug.  There is still a bug I'm having on pages with multiple iframes, which I want to figure out, and I want to make the UI nicer, and a few other things.

Thanks!
